js/editor-ta.js

```diff
@@ -218,72 +246,134 @@ EditorTextArea.prototype.onChange = function() {
   });
 };
 
+EditorTextArea.prototype.getTotalHeight = function() {
+  return this.totalCalculatedHeight_;
+};
+
+EditorTextArea.prototype.getTotalWidth = function() {
+  if (this.settings_.get('wraplines')) {
+    return this.dimentions.width;
+  } else {
+    return this.characterDimentions_.width * this.longestLine_;
+  }
+}
+
+EditorTextArea.prototype.getLineHeight = function(line) {
+  const charWidth = this.characterDimentions_.width;
+  const charHeight = this.characterDimentions_.height;
+  if (!this.settings_.get('wraplines')) {
+    return charHeight;
+  }
+  // This line box filling algorithm defines how big a line will be by
+  // determining the points at which the line will break. This is at a
+  // whitespace character, in the middle of a word if it's the only word on the
+  // line or at a line break suggestion character (see below).
+  // https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens#Suggesting_line_break_opportunities
+  const words = [];
+  const tokens = line.split(/\s/);
+  let i = 0;
+  for (let token of tokens) {
+    if (/[-\u00AD]/.test(token)) {
+      // Split a word with a line break suggestion character into chunks so each
+      // component can correctly wrap.
+      const suggestedChunks = token.split(/[-\u00AD]/);
+      let i = 0;
+      for (word of suggestedChunks) {
+        if (i < suggestedChunks.length-1) {
+          word += '-';
+        }
+        words.push(word);
+        i++;
+      }
+    } else if (token.length > 0) {
+      // If we are not at the end, add in the trailing space to the word.
+      if (i < tokens.length - 1) {
+        token += ' ';
+      }
+      words.push(token);
+    } else {
+      // If we have a empty string it means we split on 2 white space chars
+      if (i < tokens.length - 1) {
+        words.push(' ');
+      }
+    }
+    i++;
+  }
+
+  let lines = 1;
+
+  let width = this.dimentions.width;
```

Comment:
not sure what 'width' represents exactly here. the width of the textarea reamining on this line available to place words into? if so a bit more specificity in the name would be good. maybe lineWidthAvailable?